src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -191,12 +191,49 @@ impl<'a, 'gcx, 'tcx> Borrows<'a, 'gcx, 'tcx> {
         }
     }
 
-    fn kill_borrows_on_local(&self,
-                             sets: &mut BlockSets<BorrowIndex>,
-                             local: &rustc::mir::Local)
-    {
-        if let Some(borrow_indexes) = self.borrow_set.local_map.get(local) {
-            sets.kill_all(borrow_indexes);
+    /// Kill any borrows that conflict with `place`.
+    fn kill_borrows_on_place(
+        &self,
+        sets: &mut BlockSets<BorrowIndex>,
+        location: Location,
+        place: &Place<'tcx>
+    ) {
+        debug!("kill_borrows_on_place: location={:?} place={:?}", location, place);
+        // Handle the `Place::Local(..)` case first and exit early.
+        if let Place::Local(local) = place {
+            if let Some(borrow_indexes) = self.borrow_set.local_map.get(&local) {
+                debug!(
+                    "kill_borrows_on_place: local={:?} borrow_indexes={:?}",
+                    local, borrow_indexes,
+                );
+                sets.kill_all(borrow_indexes);
+                return;
+            }
+        }
+
+        // Otherwise, look at all borrows that are live and if they conflict with the assignment
+        // into our place then we can kill them.
+        let mut borrows = sets.on_entry.clone();
+        let _ = borrows.union(sets.gen_set);
+        for borrow_index in borrows.iter() {
+            let borrow_data = &self.borrows()[borrow_index];
+            debug!(
+                "kill_borrows_on_place: borrow_index={:?} borrow_data={:?}",
+                borrow_index, borrow_data,
+            );
+            if places_conflict::places_conflict(
+                self.tcx,
+                self.mir,
+                place,
+                &borrow_data.borrowed_place,
+                places_conflict::PlaceConflictBias::NoOverlap,
```

Comment:
I am still working through the overall logic here, bear with me. I think there is an underlying double-negative here that makes it all make sense, but it might warrant a comment or two.

Kill borrows on place itself does not take a `bias` argument. It just always passes `NoOverlap`, which I earlier thought was stating that any pair of indices being compared are not equal (but now am rethinking that interpretation, given how it is used here).

Of course we cannot know that the pair of indices being compared are not equal, because this function does not have enough context to determine that.

What we (maybe) can say is that when doing the recursion, the conservative option (as in the sound choice) is to assume they are unequal.

So maybe I'd add a note here, above this use of `NoOverlap`, saying something like "we conservatively assume any given pair of array indices are unequal, so that when `places_conflict` returns true, we will be assured that the two places being compared each definitely denote the same sets of locations."

Reply:
And now that I've written all that, I think I better understand why you chose the word "bias." The bias here reflects "how should I interpret the unknowable (when comparing runtime array indices)", and the answer is always derived from "what would my caller want me to do in order to make the conservative choice to ensure soundness."